scripts/code_review_tool_evaluator.py

```diff
@@ -29,6 +31,189 @@
+        - No explanation is needed in the output, only the IDs of the matched comments.
+        - The output should be a valid json only.
+
+
+**Output example**:
+
+    [
+        {{"old_comment_id": 1, "new_comment_id": 3}},
+        {{"old_comment_id": 4, "new_comment_id": 2}},
+    ]
+
+**First set of comments (old comments)**:
+
+{old_comments}
+
+**Second set of comments (new comments)**:
+
+{new_comments}
+"""
+
+
+class FeedbackEvaluator:
+    def __init__(self, evaluation_dataset: str):
+        self.evaluated_comments = pd.read_csv(evaluation_dataset)
+
+        llm = generative_model_tool.create_openai_llm()
+        evaluate_comments_prompt = PromptTemplate.from_template(EVALUATION_TEMPLATE)
+        self.evaluation_chain = evaluate_comments_prompt | llm
+
+    def evaluate_diff_comments(
+        self,
+        diff_id: int,
+        new_comments: list[code_review.InlineComment],
+    ) -> list[dict]:
+        diff_evaluated_comments = self.evaluated_comments[
+            self.evaluated_comments["diff_id"] == diff_id
+        ].reset_index()
+        diff_evaluated_comments["evaluation"] = np.where(
+            diff_evaluated_comments["evaluation"].isin(["CORRECT", "VALID_REDUNDANT"]),
+            "VALID",
+            "INVALID",
+        )
+
+        output = self.evaluation_chain.invoke(
+            {
+                "old_comments": [
+                    {
+                        "id": i,
+                        "content": raw["comment"],
+                        "file": raw["file_path"],
+                    }
+                    for i, raw in diff_evaluated_comments.iterrows()
+                ],
+                "new_comments": [
+                    {
+                        "id": i,
+                        "content": comment.content,
+                        "file": comment.filename,
+                    }
+                    for i, comment in enumerate(new_comments)
+                ],
+            }
+        )
+
+        matches = code_review.parse_model_output(output.content)
+
+        results = [
+            {
+                "new_comment": comment.content,
+                "old_comments_count": 0,
+                "matched": False,
+            }
+            for comment in new_comments
+        ]
+        seen_old_comments = set()
+
+        for match in matches:
+            old_index = match["old_comment_id"]
+            new_index = match["new_comment_id"]
+
+            evaluated_comment = diff_evaluated_comments.iloc[old_index]
+            new_comment = new_comments[new_index]
+
+            if evaluated_comment["file_path"] != new_comment.filename:
+                print(
+                    "File mismatch:",
+                    evaluated_comment["file_path"],
+                    new_comment.filename,
+                )
+                continue
+
+            current_result = results[new_index]
+
+            current_result["evaluation"] = (
+                "MIXED"
```

Comment:
When can this happen? A new comment matched to multiple old comments?

Reply:
This happens when it matches two different old comments; one of them is valid, and the other is invalid.

Reply:
Have you seen it happen frequently? It's strange to match two old comments, isn't it?

Reply:
No, it did not happen so far, but it happened; we can see it in the results.